### DISCUSSION TITLE
Organising larger applications

### DISCUSSION THREAD

**Discussion by aminalaee** [Kludex/mangum#232] [Q&A]
Hey,
First off I wanted to thank you for the great library and the effort behind it.

This might be more of a question than an issue,
Based on the examples I've seen in the docs we pass the whole application inside the `Mangum` adapter. Like:

```python
from fastapi import FastAPI
from mangum import Mangum

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}

handler = Mangum(app)
```

Well this is one convention and works for some projects. For larger projects it makes sense to split the backend into services/functions and let the Lambda scale each separately as explained  in the AWS docs [here](https://aws.amazon.com/blogs/compute/best-practices-for-organizing-larger-serverless-applications/).

I think this should be straightforward with Starlette `Router` or FastAPI `APIRouter` instances which are basically mini ASGI applications.
So maybe something like this should work out of the box:

```python
# One handler per router
from fastapi import APIRouter
from mangum import Mangum

router = APIRouter()

@router.get("/")
def read_root():
    return {"Hello": "World"}

handler = Mangum(router)
```

And another handler for another router.
So my question is if this makes sense or I am missing something?
And if this is supported maybe there can be examples or tests demonstrating this?

**Comment by jordaneremieff** [comment]
Hi @aminalaee,

To my knowledge the `APIRouter` class in `fastapi` subclassess the `Router` in `starlette`. Based on [this](https://github.com/encode/starlette/blob/master/starlette/routing.py#L641), it would appear that this class exposes an ASGI-compatible interface ([see also](https://mangum.io/asgi-frameworks/#background)), so the usage you are suggesting should work.

There is currently an open [issue](https://github.com/jordaneremieff/mangum/issues/104) related to improving the documentation and adding more examples, so I'm happy to merge any PRs that provide any framework-specific examples demonstrating things like this.

**Reply by aminalaee** [reply]
> To my knowledge the `APIRouter` class in `fastapi` subclassess the `Router` in `starlette`. Based on [this](https://github.com/encode/starlette/blob/master/starlette/routing.py#L641), it would appear that this class exposes an ASGI-compatible interface ([see also](https://mangum.io/asgi-frameworks/#background)), so the usage you are suggesting should work.

Yes, as far as I know this should be ok.

> There is currently an open [issue](https://github.com/jordaneremieff/mangum/issues/104) related to improving the documentation and adding more examples, so I'm happy to merge any PRs that provide any framework-specific examples demonstrating things like this.

Great, I'll add some docs so we can discuss this.

**Reply by aminalaee** [reply]
I have written a small blog about this https://aminalaee.dev/posts/2022/fastapi-aws-lambda/#larger-applications if this is useful I can add the related info to Mangum docs too. Please let me know @jordaneremieff

**Reply by jordaneremieff** [reply]
@aminalaee that's great, I think it would be a useful addition to https://mangum.io/external-links/ if you would like to PR a link and brief description.

**Comment by antoine-mle** [comment]
Hi @aminalaee, did you manage to do tes same process but with docker instead of zip? 
If so, could you share the process?
Best regards, 
A.
